### PR TITLE
perf: Do not retry Snuba Query on ReadTimeout

### PR DESCRIPTION
**Pull request by fpacifici**
urllib3 consider any timeout after the connection is successfully established as a read error, thus the same retry policy we set up for read error applies.
The consequence is that, when a Snuba query times out we retry. This creates additional useless load on Snuba for no reason. When snuba times out it may mean that clickhouse is overloaded or that we are reducing the number of threads to run a query. Retrying makes the situation worse. Also the client has given up by the second attempt.

Since there is no way to separate read timeout from read errors by config, this PR subclasses the Retry class from urllib3 to change the policy and adds a test to ensure this change remains relevant in case urllib3 changes.

Tested by introducing a delay in Snuba and observed that sentry gives up immediately after the first timeout while other read errors still cause retries

### CODE REVIEW

File: tests/sentry/utils/test_urllib3_timeout.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def _new_conn(self):
+        return self.connection
+
+
+def test_retries():
+    """
+    Tests that, even if I set up 5 retries, there is only one request
+    made since it times out.
+    """
+    conneciton_mock = mock.Mock()
```
Comment (tkaemming) [inline tests/sentry/utils/test_urllib3_timeout.py]:
```suggestion
    connection_mock = mock.Mock()
```

**Review by tkaemming — APPROVED** [review]

File: tests/sentry/utils/test_urllib3_timeout.py (hunk 1 above)
Comment (fpacifici) [reply tests/sentry/utils/test_urllib3_timeout.py]: I need to find a way to formally introduce the word `connecIton` in the english vocabulary

**Comment by mattrobenolt** [comment]
Just for context, I looked into other ways, and this seems the least evil of them to do this. I think we should try to upstream some behavior here since this seems like a common thing people would want.

**Comment by fpacifici** [comment]
Yea we should do that. I saw a couple of issues filed on urllib3 asking for changes in the behavior of Retry (unrelated to this problem) and the answer was generally: "subclass Retry". But this seems generic enough to try to send a PR.

**Pull request merged**